biblemulticonverter/src/test/java/biblemulticonverter/format/paratext/USX3Test.java

```diff
@@ -0,0 +1,148 @@
+		dump(usx3.doImportBook(resultFile), actual);
+
+		assertEqualLines(expected, actual);
+	}
+
+	@Test
+	public void test_USX2_end_milestone_insertion() throws Exception {
+		// 1. Import from USX 2, the USX 2 importer should add chapter and verse end milestones to the internal
+		//    format.
+		// 2. Assert that when imported the result is equal to the same file but in USX 3 format (without any USX 3
+		//    features except for the end milestones)
+
+		File originalFile = getResource("/usx2/GEN.usx");
+		File resultFile = createTempFile("usx3-export", ".usx");
+
+		USX usx = new USX();
+		ParatextBook paratextBook = usx.doImportBook(originalFile);
+		assertNotNull(paratextBook);
+
+		USX3 usx3 = new USX3();
+		usx3.doExportBook(paratextBook, resultFile);
+
+		// Compare the two files by importing them and dumping them so they can easily be compared
+		File expected = createTempFile("expected", ".txt");
+		File actual = createTempFile("actual", ".txt");
+
+		dump(usx3.doImportBook(getResource("/usx3/GEN-with-only-USX3-end-milestones.usx")), expected);
+		dump(usx3.doImportBook(resultFile), actual);
+
+		assertEqualLines(expected, actual);
+	}
+
+	@Test
+	public void test_import_USX3_export_USX() throws Exception {
+		File originalFile = getResource("/usx3/GEN-with-only-USX3-end-milestones.usx");
+		File resultFile = createTempFile("usx-export", ".usx");
+
+		USX3 usx3 = new USX3();
+		ParatextBook paratextBook = usx3.doImportBook(originalFile);
+
+		USX usx = new USX();
+		usx.doExportBook(paratextBook, resultFile);
+	}
+
+	@Test
+	public void test_import_USX3_export_USFM2() throws Exception {
+		File originalFile = getResource("/usx3/GEN-with-only-USX3-end-milestones.usx");
+		File resultFile = createTempFile("usfm-export", ".usfm");
+
+		USX3 usx3 = new USX3();
+		ParatextBook paratextBook = usx3.doImportBook(originalFile);
+
+		USFM usfm = new USFM();
+		usfm.doExportBook(paratextBook, resultFile);
+	}
+
+	private static void dump(ParatextBook book, File file) throws IOException {
+		ParatextDump paratextDump = new ParatextDump();
+		paratextDump.doExportBook(book, file);
+	}
+
+	private static void assertEqualLines(File expected, File actual) throws IOException {
+		Iterator<String> expectedLines = Files.readAllLines(expected.toPath()).iterator();
+		Iterator<String> actualLines = Files.readAllLines(actual.toPath()).iterator();
+
+		while (expectedLines.hasNext() && actualLines.hasNext()) {
+			String expectedLine = expectedLines.next();
+			String actualLine = actualLines.next();
+			assertEquals(expectedLine, actualLine);
+		}
+		assertEquals(expectedLines.hasNext(), actualLines.hasNext());
+	}
+
+	private static File createTempFile(String prefix, String suffix) throws IOException {
+		File tempFile = File.createTempFile(prefix, suffix);
+		tempFile.deleteOnExit();
+		return tempFile;
+	}
+
+	private static File getResource(String path) {
+		try {
+			return new File(USX3Test.class.getResource(path).toURI());
```

Comment:
I believe this will fail on some CI systems that run the unit tests from inside the JAR file. In that case getResource() will return a `jar:` URI which you cannot convert to a File.

Reply:
Maybe we need import and export methods that accept InputStreams/OutputStreams? So we can easily work with resources?